### ISSUE TITLE
Getting a nusty error - Screen.Authenticator.js:42 Uncaught TypeError: Cannot read property '0' of undefined

### ISSUE THREAD

**Issue by telostia** [pheryx/node-phaser-rtms#1]
Hi, I've got a problem with the error:
Screen.Authenticator.js:42 Uncaught TypeError: Cannot read property '0' of undefined
at Screen.Authenticator.initDataCreate (Screen.Authenticator.js:42)
    at Screen.Authenticator.create (Screen.Authenticator.js:23)
    at c.StateManager.loadComplete (phaser.js:30867)
    at c.StateManager.preUpdate (phaser.js:30634)
    at c.Game.updateLogic (phaser.js:39002)
    at c.Game.update (phaser.js:38893)
    at c.RequestAnimationFrame.updateRAF (phaser.js:61958)
    at window.requestAnimationFrame.forceSetTimeOut._onLoop (phaser.js:61942)

I've upgraded most of the nodejs package because i'm using a later version of nodejs 6.11.3
Heres are the package.json that I use:

![image](https://user-images.githubusercontent.com/1754262/36243273-d8cb0054-1274-11e8-999f-addfe41b6f41.png)

I've made some changes to the index.js so it reflects the updated packages

Do I need to populate the mongodb database first? it's getting a null value at line in file Screen.Authenticator.js: 

var providerKey = profile.provider + 'Id';
	console.log(' ____ '+ providerKey);

Please help.
 I'm trying to create a similar game online with a turn based battle system mmo and this would help greatly!

Dan.

**Comment by telostia** [comment]
I uploaded the git to my fork so you can try it out and see the errors more clearly.
https://github.com/telostia/node-phaser-rtms

thanks.

**Comment by pheryx** [comment]
It looks like you're getting an error here on this line:

`var playerCrit = { profileEmail: profile.emails[0].value };`

It cannot find property 0 of undefined, which means "emails" is undefined in the profile object. This project is old code, so it's probably that the profile returned from the authentication provider is a different structure where the value "emails" no longer exists.

My suggestion would be trap the profile object that gets returned from the provider, print it for debugging, and find where the new "email" field lives.

As for MongoDB, you do not need to pre-populate the database with anything, but I do believe you need the mongo daemon to be running with a database available.

**Comment by telostia** [comment]
Thanks for the reply! I understand that the criteria has been changed but I can not wrap my head around your code to make the changes. I see that on another example they did it like this:

passport.use(new GoogleStrategy({  
    clientID: configAuth.googleAuth.clientID,
    clientSecret: configAuth.googleAuth.clientSecret,
    callbackURL: configAuth.googleAuth.callbackURL,
  },
    function(token, refreshToken, profile, done) {
      process.nextTick(function() {
        User.findOne({ 'google.id': profile.id }, function(err, user) {
          if (err)
            return done(err);
          if (user) {
            return done(null, user);
          } else {
            var newUser = new User();
            newUser.google.id = profile.id;
            newUser.google.token = token;
            newUser.google.name = profile.displayName;
            newUser.google.email = profile.emails[0].value;
            newUser.save(function(err) {
              if (err)
                throw err;
              return done(null, newUser);
            });
          }
        });
      });
    }));

};

The official passport google instruction shows as:
var GoogleStrategy = require('passport-google-oauth').OAuth2Strategy;

passport.use(new GoogleStrategy({
    clientID: GOOGLE_CLIENT_ID,
    clientSecret: GOOGLE_CLIENT_SECRET,
    callbackURL: "http://127.0.0.1:3000/auth/google/callback"
  },
  function(accessToken, refreshToken, profile, done) {
    User.findOrCreate({ googleId: profile.id }, function (err, user) {
      return done(err, user);
    });
  }
));

whereas in your code it shows:
passport.use(new GoogleStrategy({
		clientID: '565648675110-fon23jen3v151p3s3efm0pmo9i24vr2e.apps.googleusercontent.com',
		clientSecret: 'aYKUIr992BdJ9mXMtS90ebGJ',
		callbackURL: 'http://localhost:4004/auth/google/callback'
	},
	function (accessToken, refreshToken, profile, callback) {
	console.log('--- profile object = ' + profile);
		callback(null, profile);
	}
));

Where and how do I get the profile object return from to print out the values that I need. Sorry I am quite new to this nodejs stuff and I'm trying to wrap my head around it.

**Comment by pheryx** [comment]
If you've upgraded the Passport module, you should probably use their examples. You may need to rework much of this code. Again, this project is several years old, and outdated. It's non-trivial to upgrade Node modules for current versions, and I don't have a simple answer, because you will likely run into more errors after fixing this one.

Did you set up your Google Developer profile and create a client app? I assume you probably covered that, but for OAuth to work with providers, you need to configure an app on their end, which should grant your Client ID and Secret, and allow you to create an authorized callback URL which will return the result to your node server, where, if successful, you can console.log the profile.

Another suggestion would be to try to remove OAuth entirely, and don't use authentication until you've repaired the code enough, then refactor it in later. Sorry I can't help more, but I too learned Node the same way, through trial and error, and am still learning.

For your game, maybe look into this project as well: https://github.com/benas/gamehub.io

It does not include OAuth, but you could add it, and the server is well-designed to support the turn-based system you mentioned. Although it appears to be a web app, you can still use it as a standalone server from your game framework. I'm using it currently for one of my game projects.

**Comment by telostia** [comment]
To answer your question, I did create a client id and secret. I only omit it due to confidential reasons and that is the reason why I replaced it with client_id and client_secret on the above example.
Actually the focus of this app is the login of users(especially social connection like facebook,google etc through which is very hard to find examples on as most of the examples aren't focused on gaming especially when applying to phaser which I am using but just web apps in general). I have a working game with phaser multiplayer applied but does not have a refined login system like you have in this code. It is using sessions to store data and I would like to improvise on it hence when I saw your code it was the perfect match except that I have to get it to work first :).  
Again sorry for the disturbance and thanks for directing me to solve this issue. regarding https://github.com/benas/gamehub.io I actually looked into that but didn't put much focus on it. I'll try to look into it more.

you mentioned:
'but I too learned Node the same way, through trial and error, and am still learning.'

Perhaps that's the only way out to fully grasps the concept and to avoid further headaches down the road of not knowing the system well, I might end up with more problems if i don't fully understand it.

Once again, this code has been a great help to me so I appreciate you placed it for everyone to use, I'll try to hack my way through it and if I somehow succeed I'll post a pull request so it could help others .
